### ISSUE TITLE
SpriteBatch for Particles

### ISSUE THREAD

**Issue by tomspilman** [MonoGame/MonoGame#5348]
So I've seen some games use `SpriteBatch` for particles at times.  While this works it is inefficient because generally the code looks like this:
```
foreach (var e in myemitters)
{
    spriteBatch.Begin();
    foreach (var p in e.particles)
       spriteBatch.Draw(e.Texture, p.Position, e.SourceRect, p.Color, p.Rotation, GetOrigin(p), p.Scale, p.FlipEffects, p.Depth);
    spriteBatch.End();
}
```
This is arguably the most expensive version of `Draw` and many redundant calculations occur for each draw.

For sure #5347 will help.

Also you could argue that people shouldn't do this... write your own rendering using a vb/drawprim that is customized to your particle system design.

But it made me think we could add some additional methods to `SpriteBatch` that help reduce redundancy when lots of related draws are made.  First and foremost when the texture/srcRect is shared which is common when you are drawing many instances of the same thing which happens a lot in games.

One approach would be to add a default texture and source rect to begin:
```
SpriteBatch.Begin(Texture2D, Rectangle, SpriteSortMode, BlendState, SamplerState, DepthStencilState, RasterizerState, Effect, Matrix);
```
Maybe a this begin call should have some alternate name.

Then add additional draw methods that don't require any texture info and instead use the pre-calculated defaults.

**Comment by tomspilman** [comment]
Another consideration is having additional `Draw` methods that pass the UVs for directly instead of letting `SpriteBatch` recalculate them from the texture and source rect.

**Comment by prime31** [comment]
I like the idea. I personally went with your second option ([drawRaw example](https://github.com/prime31/nez/blob/master/Nez.Portable/Graphics/Batcher/Batcher.cs#L462)) when implementing this. Basically a Draw that lets you pass in the exact values set for a given quad.

**Comment by willmotil** [comment]
> Then add additional draw methods that don't require any texture info and instead use the pre-calculated defaults.

This would be a great help i think for the whole framework

Pushing a texture draw location and texture uv then having the api do the drawing more efficiently under the hood ! While at the same time making it easier to use !
Would be very cool i dunno how you would name it or organize it. 
I imagine something like this that could be continually optimized under the hood.

PushedSprite(Texture2D t, Rectangle destination, Rectangle source , ... , ... , ... );

PushedSprite.Source(Rectangle sourceUv);
PushedSprite.Destination(Rectangle destinationPos);

PushedSprite.Draw();

PushedSprite.DrawTo(Rectangle destinationPos);
PushedSprite.DrawTo(Rectangle destinationPos, Rectangle sourceUv);
//PushedSprite.DrawFrom( Rectangle sourceUv); ?

**Comment by Thraka** [comment]
My [engine](https://github.com/Thraka/SadConsole) is hugely SpriteBatch heavy. Any perf improvements here would be great. I've ported the engine to SFML and doing perf tests between the two backends and pushing 16,000 sprites to the screen each frame has these results:

| Drawing Library | FPS |
| --- | --- |
| MonoGame DX | ~400 |
| MonoGame GL | ~400 |
| SFML | ~900 |

Now with SFML there is no "SpriteBatch" type class, so I had to make my own. It doesn't come anywhere near the complexity of the MonoGame SpriteBatch. It just builds a bunch of quad draws into buffers each frame and renders.

Anything that gives me better perf on MonoGame would be welcome. Even a different class that was just a more optimized version of SpriteBatch would be fine with me.

**Comment by tomspilman** [comment]
>  It just builds a bunch of quad draws into buffers each frame and renders.

Rolling your own sprite batching which is custom tailored to your needs should in theory always be faster than a general system like `SpriteBatch`.

Still our goal should be to immunize that difference and I think we can do that with the ideas in this thread.

**Comment by Thraka** [comment]
Agreed. SpriteBatch does work, and it's easy to use. I would rather use framework-based and proven code than making my own. So any improvements you do in this area would make me happy.

**Comment by ala53** [comment]
Back when I was working on a game, I found SpriteBatch was far too slow for particles. The best solution was to compute the transforms on the GPU -- memory bandwidth to the GPU was far less limiting than CPU performance. It ended up being about 10-15x faster than naive SpriteBatch calls on a laptop with a ULV processor.

Anyhow, it might be worth it to use a special shader for SpriteBatch calls and do all the work on the GPU (even though it uses much more memory per vertex).

See: 
https://github.com/ala53/mptanks2d/blob/develop/MPTanks-MK5/Client/Backend/Renderer/PreProcessorTypes/ParticlePreProcessor.cs#L30
https://github.com/ala53/mptanks2d/blob/develop/MPTanks-MK5/Client/Backend/Renderer/LayerRenderers/GameWorldRenderer.cs#L168
https://github.com/ala53/mptanks2d/blob/develop/MPTanks-MK5/Client/Backend/assets/mgcontent/componentRenderer.fx
